common/src/types/v0/message_bus/volume.rs

```diff
@@ -8,45 +9,91 @@ bus_impl_string_uuid!(VolumeId, "UUID of a mayastor volume");
 /// Volumes
 ///
 /// Volume information
-#[derive(Serialize, Deserialize, Default, Debug, Clone, Eq, PartialEq)]
+#[derive(Serialize, Deserialize, Default, Debug, Clone, PartialEq)]
 #[serde(rename_all = "camelCase")]
 pub struct Volume {
+    /// Desired specification of the volume.
+    spec: VolumeSpec,
+    /// Runtime state of the volume.
+    state: Option<VolumeState>,
+}
+
+impl Volume {
+    /// Construct a new volume.
+    pub fn new(spec: &VolumeSpec, state: &Option<VolumeState>) -> Self {
+        Self {
+            spec: spec.clone(),
+            state: state.clone(),
+        }
+    }
+
+    /// Get the volume spec.
+    pub fn get_spec(&self) -> VolumeSpec {
+        self.spec.clone()
+    }
+
+    /// Get the volume state.
+    pub fn get_state(&self) -> Option<VolumeState> {
+        self.state.clone()
+    }
+}
+
+impl From<Volume> for models::Volume {
+    fn from(volume: Volume) -> Self {
+        Self {
+            spec: volume.get_spec().into(),
+            state: volume.state.map(|state| state.into()),
+        }
+    }
+}
+
+/// Runtime volume state information.
+#[derive(Serialize, Deserialize, Default, Debug, Clone, Eq, PartialEq)]
+#[serde(rename_all = "camelCase")]
+pub struct VolumeState {
     /// name of the volume
     pub uuid: VolumeId,
     /// size of the volume in bytes
     pub size: u64,
-    /// current state of the volume
-    pub state: VolumeState,
+    /// current status of the volume
+    pub status: VolumeStatus,
     /// current share protocol
     pub protocol: Protocol,
     /// array of children nexuses
     pub children: Vec<Nexus>,
 }
 
-impl From<Volume> for models::Volume {
-    fn from(src: Volume) -> Self {
-        Self::new(
-            src.children,
-            src.protocol,
-            src.size,
-            src.state,
-            apis::Uuid::try_from(src.uuid).unwrap(),
-        )
+impl From<VolumeState> for models::VolumeState {
+    fn from(volume: VolumeState) -> Self {
+        Self {
+            children: volume.children.iter().cloned().map(|c| c.into()).collect(),
```

Comment:
volume.children.into_vec() ?

Reply:
Thanks, I was looking for that and couldn't find it.